langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java

```diff
@@ -0,0 +1,244 @@
+     * @param distanceType: the optional distanceType (default: "cosine")
+     * @param metadataPrefix: the optional metadata prefix (default: "")
+     * @param text: the optional text property name (default: "text")
+     * @param indexName: the optional index name (default: "langchain-embedding-index")
+     */
+    @Builder
+    public Neo4jEmbeddingStore(
+            SessionConfig config,
+            Driver driver,
+            int dimension,
+            String label,
+            String property,
+            Neo4jDistanceType distanceType,
+            String metadataPrefix,
+            String text,
+            String indexName) {
+        
+        /* required configs */
+        ensureNotNull(driver, "driver");
+        ensureNotNull(dimension, "dimension");
+
+        this.driver = driver;
+        this.dimension = dimension;
+        
+        /* optional configs */
+        this.config = getOrDefault(config, SessionConfig.forDatabase("neo4j"));
+        this.label = getOrDefault(label, DEFAULT_LABEL);
+        this.embeddingProperty = getOrDefault(property, DEFAULT_EMBEDDING_PROP);
+        this.distanceType = getOrDefault(distanceType, Neo4jDistanceType.COSINE);
+        this.indexName = getOrDefault(indexName, DEFAULT_IDX_NAME);
+        this.metadataPrefix = getOrDefault(metadataPrefix, "");
+        this.text = getOrDefault(text, DEFAULT_TEXT_PROP);
+        
+        /* auto-index creation */
+        createIndexIfNotExist();
+    }
+
+    /*
+    Methods with `@Override`
+    */
+    
+    @Override
+    public String add(Embedding embedding) {
+        String id = randomUUID();
+        add(id, embedding);
+        return id;
+    }
+
+    @Override
+    public void add(String id, Embedding embedding) {
+        addInternal(id, embedding, null);
+    }
+
+    @Override
+    public String add(Embedding embedding, TextSegment textSegment) {
+        String id = randomUUID();
+        addInternal(id, embedding, textSegment);
+        return id;
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings) {
+        return addAll(embeddings, null);
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedded) {
+        List<String> ids = embeddings.stream()
+                .map(ignored -> randomUUID())
+                .toList();
+        addAllInternal(ids, embeddings, embedded);
+        return ids;
+    }
+
+    @Override
+    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
+        var embeddingValue = Values.value(referenceEmbedding.vector());
+        
+        try (var session = session()) {
+            return session
+                    .run("""
+						CALL db.index.vector.queryNodes($indexName, $maxResults, $embeddingValue)
+						YIELD node, score
+						WHERE score >= $minScore
+						RETURN node, score
```

Comment:
can we make configurable which properties we want to return or the full node?
because if we return all the huge properties like the embedding + big texts that can become a lot

we should also add the option like in langchain integration to provide a retrievalQuery in the config that is applied after the node/score is retrieved from the index so that the user can optionally find more contextually relevant data in the neighbourhood.

Reply:
Added a retrievalQuery config, with default:
```
RETURN properties(node) AS metadata, node.id AS id, node.text AS text, node.embedding AS embedding, score
```